access/jira/app.go

```diff
@@ -297,82 +338,250 @@ func (a *App) onJIRAWebhook(ctx context.Context, webhook Webhook) error {
 		return trace.Wrap(err)
 	}
 	log.Infof("JIRA user %s the request", resolution)
 
 	return nil
 }
 
-func (a *App) onPendingRequest(ctx context.Context, req access.Request) error {
-	reqData := RequestData{User: req.User, Roles: req.Roles, RequestReason: req.RequestReason, Created: req.Created}
-	jiraData, err := a.bot.CreateIssue(ctx, req.ID, reqData)
+func (a *App) onPendingRequest(ctx context.Context, req types.AccessRequest) error {
+	reqID := req.GetName()
+	reqData := RequestData{
+		User:          req.GetUser(),
+		Roles:         req.GetRoles(),
+		RequestReason: req.GetRequestReason(),
+		Created:       req.GetCreationTime(),
+	}
 
+	isNew, err := a.modifyPluginData(ctx, reqID, func(existing *PluginData) (PluginData, bool) {
+		if existing != nil {
+			return PluginData{}, false
+		}
+		return PluginData{RequestData: reqData}, true
+	})
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	logger.Get(ctx).WithFields(logger.Fields{
-		"jira_issue_id":  jiraData.ID,
-		"jira_issue_key": jiraData.Key,
-	}).Info("JIRA Issue created")
+	if isNew {
+		if err := a.createIssue(ctx, reqID, reqData); err != nil {
+			return trace.Wrap(err)
+		}
+	}
 
-	err = a.setPluginData(ctx, req.ID, PluginData{reqData, jiraData})
+	if reqReviews := req.GetReviews(); len(reqReviews) > 0 {
+		if err = a.addReviewComments(ctx, reqID, reqReviews); err != nil {
+			return trace.Wrap(err)
+		}
+	}
 
-	return trace.Wrap(err)
+	return nil
 }
 
-func (a *App) onDeletedRequest(ctx context.Context, req access.Request) error {
-	log := logger.Get(ctx)
-	reqID := req.ID // This is the only available field
+func (a *App) onResolvedRequest(ctx context.Context, req types.AccessRequest) error {
+	var commentErr error
+	if err := a.addReviewComments(ctx, req.GetName(), req.GetReviews()); err != nil {
+		commentErr = trace.Wrap(err)
+	}
 
-	pluginData, err := a.getPluginData(ctx, reqID)
+	resolution := Resolution{Reason: req.GetResolveReason()}
+	switch req.GetState() {
+	case types.RequestState_APPROVED:
+		resolution.Tag = ResolvedApproved
+	case types.RequestState_DENIED:
+		resolution.Tag = ResolvedDenied
+	}
+	err := trace.Wrap(a.resolveIssue(ctx, req.GetName(), resolution))
+	return trace.NewAggregate(commentErr, err)
+}
+
+func (a *App) onDeletedRequest(ctx context.Context, reqID string) error {
+	return a.resolveIssue(ctx, reqID, Resolution{Tag: ResolvedExpired})
+}
+
+// createIssue posts a JIRA issue with request information.
+func (a *App) createIssue(ctx context.Context, reqID string, reqData RequestData) error {
+	data, err := a.jira.CreateIssue(ctx, reqID, reqData)
 	if err != nil {
-		if trace.IsNotFound(err) {
-			log.WithError(err).Warn("Cannot expire unknown request")
-			return nil
+		return trace.Wrap(err)
+	}
+
+	ctx, log := logger.WithFields(ctx, logger.Fields{
+		"jira_issue_id":  data.IssueID,
+		"jira_issue_key": data.IssueKey,
+	})
+	log.Info("JIRA Issue created")
+
+	_, err = a.modifyPluginData(ctx, reqID, func(existing *PluginData) (PluginData, bool) {
```

Comment:
Same comment as for Pagerduty - https://github.com/gravitational/teleport-plugins/pull/220#discussion_r655796923
Can you name these anonymous functions, perhaps with a comment, to make it clear what their purpose is?